gloo/transport/tcp/pair.cc

```diff
@@ -151,28 +151,90 @@ void Pair::setSync(bool sync, bool busyPoll) {
   busyPoll_ = busyPoll;
 }
 
-void Pair::listen() {
-  std::lock_guard<std::mutex> lock(m_);
+int Pair::bindToPort(int fd, unsigned short port) {
   int rv;
-
-  const auto& attr = device_->attr_;
-  auto fd = socket(attr.ai_family, attr.ai_socktype, attr.ai_protocol);
-  if (fd == -1) {
-    signalAndThrowException(GLOO_ERROR_MSG("socket: ", strerror(errno)));
-  }
-
   // Set SO_REUSEADDR to signal that reuse of the listening port is OK.
   int on = 1;
   rv = setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, &on, sizeof(on));
   if (rv == -1) {
-    ::close(fd);
     signalAndThrowException(GLOO_ERROR_MSG("setsockopt: ", strerror(errno)));
   }
 
-  rv = bind(fd, (const sockaddr*)&attr.ai_addr, attr.ai_addrlen);
-  if (rv == -1) {
-    ::close(fd);
-    signalAndThrowException(GLOO_ERROR_MSG("bind: ", strerror(errno)));
+  // Set the port we are looking for.
+  const auto& attr = device_->attr_;
+  struct sockaddr_storage bindAddr;
+  memcpy(&bindAddr, &attr.ai_addr, attr.ai_addrlen);
+  if (attr.ai_family == AF_INET) {
+    ((struct sockaddr_in*)&bindAddr)->sin_port = htons(port);
+  } else if (attr.ai_family == AF_INET6) {
+    ((struct sockaddr_in6*)&bindAddr)->sin6_port = htons(port);
+  } else {
+    GLOO_THROW_INVALID_OPERATION_EXCEPTION("unknown sa_family");
+  }
+
+  return bind(fd, (struct sockaddr*)&bindAddr, attr.ai_addrlen);
+}
+
+void Pair::listen() {
+  std::lock_guard<std::mutex> lock(m_);
+  int rv;
+  int fd;
+
+  // Check for a port range.
+  char *portRange = getenv("GLOO_PORT_RANGE");
+  unsigned short portMin, portMax;
+  bool portRangeValid = false;
+  if (portRange != NULL) {
+    errno = 0;
+    int ret = sscanf(portRange, "%hu:%hu", &portMin, &portMax);
+    if (ret == 2 && portMax >= portMin) {
+      portRangeValid = true;
+    } else {
+      signalAndThrowException(
+          GLOO_ERROR_MSG("invalid GLOO_PORT_RANGE: ", portRange));
+    }
+  }
+
+  const auto& attr = device_->attr_;
+
+  if (portRangeValid) {
+    // Find a port in the specified port range.
+    bool foundPort = false;
+    for (size_t i = portMin; i <= portMax; i++) {
```

Comment:
`unsigned short int`?

Reply:
no, because if portMax was USHRT_MAX you'd wrap to zero at the top and then bind to any port.  I could change `size_t` to `unsigned int` if you'd prefer?